Turn on -Wsign-compare

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74996

This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Turn on -Wsign-compare

This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Turn on -Wsign-compare"


This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Turn on -Wsign-compare"


This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Turn on -Wsign-compare"


This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

dagitses reviewed (commented):
looks good to me, but deferring to nikita for cmake changes.

dagitses commented on c10/util/int128.cpp:
```diff
@@ -171,7 +171,7 @@ std::ostream& operator<<(std::ostream& o, const uint128& b) {
   // Add the requisite padding.
   std::streamsize width = o.width(0);
-  if (width > rep.size()) {
+  if (static_cast<std::string::size_type>(width) > rep.size()) {
```
thanks, that's good enough for me!

malfet reviewed (approved):
Looks good to me, but I wonder whether this is already covered by https://github.com/pytorch/pytorch/pull/75085 
Also, why enable it just for C10 instead of the entire codebase?

ezyang commented:
@malfet At least one difference is that this change applies to gcc, while yours applies to clang only. I didn't enable for the whole codebase because I noticed there were a lot of existing failures and I didn't want to fix them all immediately.

ezyang commented:
@pytorchbot merge this

ezyang committed:
Update on "Turn on -Wsign-compare"


This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

malfet commented:
@ezyang I believe I've fixed all those violations(see stack ending with https://github.com/pytorch/pytorch/pull/75084), so feel free to try enabling it globally. Also, I wonder if `-Wsign-compare` should be enabled instead of its `-Werror` variant  and then respect WERROR option, which is enabled for all CI/CD builds right now.
Very likely, users with newer version of compiler will hit some violation and will have no way of suppressing those

ezyang commented:
>  Also, I wonder if -Wsign-compare should be enabled instead of its -Werror variant and then respect WERROR option, which is enabled for all CI/CD builds right now.

Hmm, I'm not sure what you mean here. How would this help with the new compiler case?

ezyang commented:
@pytorchbot merge this
